tools/linter/adapters/pyproject_linter.py

```diff
@@ -0,0 +1,277 @@
+from __future__ import annotations
+
+import argparse
+import concurrent.futures
+import json
+import logging
+import operator
+import os
+import re
+import sys
+from enum import Enum
+from pathlib import Path
+from typing import NamedTuple
+
+from packaging.version import Version
+
+
+if sys.version_info >= (3, 11):
+    import tomllib
+else:
+    import tomli as tomllib  # type: ignore[import-not-found]
+
+
+REQUIRES_PYTHON_PATTERN = re.compile(
+    r"""
+    ^\s*
+    (?P<min_ver_op>>=)
+    \s*
+    (?P<min_ver>\d+\.\d+(?:\.\d+)?)
+    \s*,\s*
+    (?P<max_ver_op><(=?))
+    \s*
+    (?P<max_ver>\d+\.\d+(?:\.\d+)?)
+    \s*$
+    """,
+    flags=re.VERBOSE,
+)
+COMPARE_OPS = {
+    "<": operator.lt,
+    "<=": operator.le,
+    ">": operator.gt,
+    ">=": operator.ge,
+    "==": operator.eq,
+    "!=": operator.ne,
+}
+
+
+class LintSeverity(str, Enum):
+    ERROR = "error"
+    WARNING = "warning"
+    ADVICE = "advice"
+    DISABLED = "disabled"
+
+
+class LintMessage(NamedTuple):
+    path: str | None
+    line: int | None
+    char: int | None
+    code: str
+    severity: LintSeverity
+    name: str
+    original: str | None
+    replacement: str | None
+    description: str | None
+
+
+def format_error_message(
+    filename: str,
+    error: Exception | None = None,
+    *,
+    message: str | None = None,
+) -> LintMessage:
+    if message is None and error is not None:
+        message = f"Failed due to {error.__class__.__name__}:\n{error}"
+    return LintMessage(
+        path=filename,
+        line=None,
+        char=None,
+        code="PYPROJECT",
+        severity=LintSeverity.ERROR,
+        name="pyproject.toml consistency",
+        original=None,
+        replacement=None,
+        description=message,
+    )
+
+
+def check_file(filename: str) -> list[LintMessage]:
+    path = Path(filename).absolute()
+    content = path.read_text(encoding="utf-8")
+    try:
+        pyproject = tomllib.loads(content)
```

Comment:
Just load it from the file descriptor / path lib file directly. No need for temporary string.